File: scripts/debug_server.py
#!/usr/bin/env python3
"""
Debug Log Server

Receives logs via HTTP POST and writes them to {project}/.claude/debug.log

Usage:
    python3 debug_server.py /path/to/project
    python3 debug_server.py  # Uses current directory
"""
from http.server import HTTPServer, BaseHTTPRequestHandler
import json
import os
import sys
from datetime import datetime

LOG_DIR = sys.argv[1] if len(sys.argv) > 1 else "."
LOG_FILE = os.path.join(LOG_DIR, ".claude", "debug.log")
PORT = int(os.environ.get("DEBUG_PORT", "8787"))

os.makedirs(os.path.dirname(LOG_FILE), exist_ok=True)


class DebugHandler(BaseHTTPRequestHandler):
    def log_message(self, *args):
        pass  # Suppress default logging

    def _send_cors_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def do_OPTIONS(self):
        self.send_response(204)
        self._send_cors_headers()
        self.end_headers()

    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self._send_cors_headers()
        self.end_headers()
        self.wfile.write(json.dumps({"status": "ok", "log_file": LOG_FILE}).encode())

    def do_POST(self):
        try:
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)
            data = json.loads(body) if body else {}

            entry = {
                "ts": datetime.now().isoformat(),
                **data
            }

            with open(LOG_FILE, "a") as f:
                f.write(json.dumps(entry) + "\n")

            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self._send_cors_headers()
            self.end_headers()
            self.wfile.write(b'{"ok":true}')

            msg = entry.get("msg", json.dumps(entry)[:80])
            print(f"[LOG] {msg}")

        except Exception as e:
            self.send_response(400)
            self.send_header("Content-Type", "application/json")
            self._send_cors_headers()
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(e)}).encode())


def main():
    print(f"Debug Log Server")
    print(f"  Endpoint: http://localhost:{PORT}/log")
    print(f"  Log file: {LOG_FILE}")
    print(f"  Press Ctrl+C to stop\n")

    server = HTTPServer(("", PORT), DebugHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nServer stopped.")


if __name__ == "__main__":
    main()

File: scripts/debug_server.js
#!/usr/bin/env node
/**
 * Debug Log Server
 *
 * Receives logs via HTTP POST and writes them to {project}/.claude/debug-{sessionId}.log
 * Supports multiple concurrent sessions via sessionId parameter.
 *
 * Usage:
 *     node debug_server.js /path/to/project
 *     node debug_server.js  # Uses current directory
 */
const http = require('http');
const fs = require('fs');
const path = require('path');

const LOG_DIR = process.argv[2] || '.';
const CLAUDE_DIR = path.join(LOG_DIR, '.claude');
const PORT = parseInt(process.env.DEBUG_PORT || '8787', 10);

fs.mkdirSync(CLAUDE_DIR, { recursive: true });

const getLogFile = (sessionId) => path.join(CLAUDE_DIR, `debug-${sessionId}.log`);

const server = http.createServer((req, res) => {
  const cors = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
  };

  if (req.method === 'OPTIONS') {
    res.writeHead(204, cors);
    res.end();
    return;
  }

  if (req.method === 'GET') {
    res.writeHead(200, { 'Content-Type': 'application/json', ...cors });
    res.end(JSON.stringify({ status: 'ok', log_dir: CLAUDE_DIR }));
    return;
  }

  if (req.method === 'POST' && req.url === '/log') {
    let body = '';
    req.on('data', chunk => (body += chunk));
    req.on('end', () => {
      try {
        const data = body ? JSON.parse(body) : {};
        const sessionId = data.sessionId || 'default';
        const logFile = getLogFile(sessionId);
        
        // Remove sessionId from stored entry (it's in filename)
        const { sessionId: _, ...rest } = data;
        const entry = { ts: new Date().toISOString(), ...rest };

        fs.appendFileSync(logFile, JSON.stringify(entry) + '\n');

        res.writeHead(200, { 'Content-Type': 'application/json', ...cors });
        res.end(JSON.stringify({ ok: true, log_file: logFile }));

        console.log(`[${sessionId}] ${entry.msg || JSON.stringify(entry).slice(0, 80)}`);
      } catch (e) {
        res.writeHead(400, { 'Content-Type': 'application/json', ...cors });
        res.end(JSON.stringify({ error: e.message }));
      }
    });
    return;
  }

  res.writeHead(404, cors);
  res.end('Not found');
});

server.listen(PORT, () => {
  console.log('Debug Log Server');
  console.log(`  Endpoint: http://localhost:${PORT}/log`);
  console.log(`  Log dir:  ${CLAUDE_DIR}`);
  console.log('  Press Ctrl+C to stop\n');
});
